File: 0765_Couples_Holding_Hands/solution.cpp
/*
765. Couples Holding Hands

https://leetcode.com/problems/couples-holding-hands/
*/

#include <string>
#include <climits>
#include <vector>
#include <unordered_map>

using namespace std;

/*
https://mli9502.github.io/2018/10/24/leetcode-765-Couples-Holding-Hands/
*/
class Solution
{
public:
  int minSwapsCouples(vector<int> &row)
  {
    unordered_map<int, int> hm;
    for (int i = 0; i < row.size(); i++)
    {
      // Assign group label to couples.
      // 0 1: 0
      // 2 3: 2
      // 4 5: 4
      // ...
      hm[row[i]] = row[i] & (~0x1);
    }

    // For each pair, if they are in two groups, we union the two groups together.
    // Every time we do a union, we have one less connected components.
    // The maximum number of connected components we can have is the number of couples, which is row.size() / 2.
    int maxNumOfConnectedComps = row.size() / 2;
    int initConnectedComps = maxNumOfConnectedComps;
    for (int i = 0; i < row.size(); i += 2)
    {
      int g1 = uf_find(hm, row[i]);
      int g2 = uf_find(hm, row[i + 1]);
      if (g1 != g2)
      {
        uf_union(hm, g1, g2);
        initConnectedComps--;
      }
    }
    // Now, initConnectedComps represents the number of connected components we have.
    // # of swaps needed = max # of connected components - # of connected components we have.
    // This is because in a given connected component, for each swap, we can at most generate one additional connected component.
    // So, # of swaps needed + # of connected components we currently have = max # of connected components.
    return maxNumOfConnectedComps - initConnectedComps;
  }
  void uf_union(unordered_map<int, int> &hm, int g1, int g2)
  {
    hm[g2] = g1;
  }
  int uf_find(unordered_map<int, int> &hm, int val)
  {
    if (hm[val] == val)
    {
      return val;
    }
    else
    {
      return uf_find(hm, hm[val]);
    }
  }
};

/*
https://blog.csdn.net/magicbean2/article/details/79628194


采用贪心的思路：为了便于快速查找某个id的人所处的位置，我们定义一个哈希表，表示从某人的id到其在row中的位置。
然后就开始扫描row中的偶数位。根据在该偶数位上的人的id，可以推出其另一半所处的位置。如果他的另一半不在他接下来的位置，
那么我们就交换下一位置和它的另一半的位置（并且记得更新哈希表）。这样最后返回总共的交换次数即可。
*/
class Solution2{
public:
    int minSwapsCouples(vector<int>& row) {
        unordered_map<int, int> hash;   // map from person id to position
        for (int i = 0; i < row.size(); ++i) {
            hash[row[i]] = i;
        }
        int ret = 0;
        for (int i = 0; i < row.size(); i += 2) {
            int p1 = row[i];
            int p2 = p1 % 2 == 0 ? p1 + 1 : p1 - 1; // its couple's id
            if (row[i + 1] != p2) {
                int p2_pos = hash[p2];
                hash[row[i + 1]] = p2_pos;
                hash[p2] = i + 1;
                swap(row[i + 1], row[p2_pos]);
                ++ret;
            }
        }
        return ret;
    }
};

File: 0765_Couples_Holding_Hands/solution.java
package cn.huang.leetcode;

/*
765. Couples Holding Hands

 N couples sit in 2N seats arranged in a row and want to hold hands. We want to know the minimum number of swaps so that every couple is sitting side by side. A swap consists of choosing any two people, then they stand up and switch seats.

The people and seats are represented by an integer from 0 to 2N-1, the couples are numbered in order, the first couple being (0, 1), the second couple being (2, 3), and so on with the last couple being (2N-2, 2N-1).

The couples' initial seating is given by row[i] being the value of the person who is initially sitting in the i-th seat.

Example 1:

Input: row = [0, 2, 1, 3]
Output: 1
Explanation: We only need to swap the second (row[1]) and third (row[2]) person.

Example 2:

Input: row = [3, 2, 0, 1]
Output: 0
Explanation: All couples are already seated side by side.

Note:

    len(row) is even and in the range of [4, 60].
    row is guaranteed to be a permutation of 0...len(row)-1.

 */
public class LC_0765_CouplesHoldingHands {
    /*
    http://www.cnblogs.com/grandyang/p/8716597.html
    ，一个数‘异或’上1就是其另一个位，如果是偶数的话，最后位是0，‘异或’上1等于加了1，变成了可以的成对奇数。
    如果是奇数的话，最后位是1，‘异或’上1后变为了0，变成了可以的成对偶数。
     */
    public int minSwapsCouples(int[] row) {
        int res = 0, n = row.length;
        for (int i = 0; i < n; i += 2) {
            if (row[i + 1] == (row[i] ^ 1)) continue;
            ++res;
            for (int j = i + 1; j < n; ++j) {
                if (row[j] == (row[i] ^ 1)) {
                    row[j] = row[i + 1];
                    row[i + 1] = row[i] ^ 1;
                    break;
                }
            }
        }
        return res;
    }

    /*
    https://blog.csdn.net/u014688145/article/details/79057475
     */
    public int minSwapsCouples2(int[] row) {
        int n = row.length;
        int ans = 0;
        for (int i = 0; i < n; i += 2) {
            int dest;
            if ((row[i] & 1) == 0)
                dest = row[i] + 1;
            else
                dest = row[i] - 1;
            if (row[i + 1] == dest)
                continue;
            ans++;
            for (int j = i + 2; j < n; ++j) {
                if (row[j] == dest) {
                    int tmp = row[i + 1];
                    row[i + 1] = row[j];
                    row[j] = tmp;
                    break;
                }
            }
        }
        return ans;
    }
}
